### PR TITLE
DYNA adapt to the new memory buffer, which has a new sample and sample_consecutive function. 

### PR DESCRIPTION
**Pull request by qiaoting159753**
Add a sample_next function into the memory_buffer.py to train an integrated world model. An integrated world model trains the reward prediction and dynamic (next state) prediction together.

Instead of 
sample function : (s, a , r, n_s, done)
sample_next function : (s, a, r, n_s, done, n_a, n_r).

### CODE REVIEW

File: cares_reinforcement_learning/memory/memory_buffer.py [hunk 1]
```diff
@@ -40,6 +40,29 @@ def sample(self, batch_size):
         transposed_batch = zip(*experience_batch)
         return transposed_batch
 
+    def sample_next(self, batch_size):
```
Comment (beardyFace) [inline cares_reinforcement_learning/memory/memory_buffer.py:43]: wonder if if there is a better name than sample_next

File: cares_reinforcement_learning/memory/memory_buffer.py [hunk 2]
```diff
@@ -40,6 +40,29 @@ def sample(self, batch_size):
...
+        """
+        For training MBRL to predict rewards. The right next transition is
```
Comment (beardyFace) [inline cares_reinforcement_learning/memory/memory_buffer.py:45]: as in correct?

File: cares_reinforcement_learning/memory/memory_buffer.py [hunk 3]
```diff
@@ -40,6 +40,29 @@ def sample(self, batch_size):
...
+        sampled as well. WHEN THE BUFFER IS NOT SHUFFLED.
+
+        (State, action, reward, next_state, next_action, next_reard)
```
Comment (beardyFace) [inline cares_reinforcement_learning/memory/memory_buffer.py:48]: how is this different to the normal buffer exactly?

File: cares_reinforcement_learning/memory/memory_buffer.py (hunk 1 above)
Comment (qiaoting159753) [reply cares_reinforcement_learning/memory/memory_buffer.py:43]: How about sample_two_steps()?

File: cares_reinforcement_learning/memory/memory_buffer.py (hunk 2 above)
Comment (qiaoting159753) [reply cares_reinforcement_learning/memory/memory_buffer.py:45]: (s1, a1, r1, s2, done) --> (s2, a2, r2, s3, done)

normal sampling: (s1, a1, r1, s2, done) 
sample_next: (s1, a1, r1, s2, done, a2, r2)

File: cares_reinforcement_learning/memory/memory_buffer.py (hunk 3 above)
Comment (qiaoting159753) [reply cares_reinforcement_learning/memory/memory_buffer.py:48]: The buffer is the same. Just this function will retrieve the next timestep's action and reward for training a world model that also predict reward.

**Review by retinfai — CHANGES_REQUESTED** [review]
I think personally you creating your own memory buffer for MBRL stuff is better. I do something similar for Hoda's RDTD3.

image

It'd solve Henry's concern about the naming. You'd just replace normal sample with your new sample. Because you don't need normal sample in your training loop (just from a quick skim from the gymnasium side)

**Comment by qiaoting159753** [comment]
> I think personally you creating your own memory buffer for MBRL stuff is better. I do something similar for Hoda's RDTD3.
> 
> image
> 
> It'd solve Henry's concern about the naming. You'd just replace normal sample with your new sample. Because you don't need normal sample in your training loop (just from a quick skim from the gymnasium side)
> 
> And so you'd import it like so:
> 
> from cares_reinforcement_learning.memory.mbrl import MemoryBuffer
> 
> You'd have to update memory_factory for the new buffer also
> 
> what do you think


Yea nah... I could do that. That is actually a good idea, for example, we can sample episodes or sample transitions for a certain length to train the network. But I don't think it will be truly worth to do it now. Because MBRL does not need a fundamental change of the memory. And let's keep it simple for now? If a new MemoryBuffer needs to inherit this memorybuffer, it can be leave as "not implemented".

**Review by retinfai — COMMENTED** [review]
sweet,that's fair

**Comment by beardyFace** [comment]
why not just extend the existing one? and override the sample method?

**Comment by qiaoting159753** [comment]
> why not just extend the existing one? and override the sample method?
> 
> The current request just duplicates the code entirely - which also isn't ideal



> why not just extend the existing one? and override the sample method?
> 
> The current request just duplicates the code entirely - which also isn't ideal

Right. Extending is a good idea. However, I think the original sample() method is still useful. Data from the original sample() is used to train the agent, like normal training. The newly implemented sample_consecutive() is much slower. It will slow down the program if sample_consecutive() is used for normal training.

**Comment by beardyFace** [comment]
I mean extend the MemoryBuffer class - not modify the existing sample function.

**Comment by qiaoting159753** [comment]
yes. I was trying to explain that there is no need to override the sample() function.

**Comment by beardyFace** [comment]
We will need to readdress this with pull request #134 - don't merge this until that is done

**Comment by beardyFace** [comment]
Update to match #140

**Review by beardyFace — APPROVED** [review]

**Pull request merged**